modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/MinimumRequiredTimeCollectorServiceImpl.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.raft;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.ignite.internal.replicator.TablePartitionId;
+
+/** Collects minimum required timestamp for each partition. */
+public class MinimumRequiredTimeCollectorServiceImpl implements MinimumRequiredTimeCollectorService {
+
+    private final ConcurrentHashMap<TablePartitionId, Long> partitions = new ConcurrentHashMap<>();
+
+    /** {@inheritDoc} */
+    @Override
+    public synchronized void addPartition(TablePartitionId tablePartitionId) {
+        partitions.put(tablePartitionId, UNDEFINED_MIN_TIME);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public synchronized void recordMinActiveTxTimestamp(TablePartitionId tablePartitionId, long timestamp) {
+        Long time = partitions.get(tablePartitionId);
+        if (time == null) {
+            // Ignore removed partitions.
+            return;
+        }
+
+        if (time == UNDEFINED_MIN_TIME || timestamp > time) {
+            partitions.put(tablePartitionId, timestamp);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void removePartition(TablePartitionId tablePartitionId) {
```

Comment:
As mentioned above, I'd rather consider different approach. However, if you will stick with given on - should we mark removePartition as synchronized? Otherwise you'll have a race with recordMinActiveTxTimestamp.

Reply:
Added `synchronised`.